dot/state/grandpa.go

```diff
@@ -26,42 +37,477 @@ var (
+				return nil, fmt.Errorf("cannot verify ancestry: %w", err)
+			}
+
+			if !isDescendant {
+				continue
+			}
+		}
+
+		// the changes must be applied in order, so we need to check if our finalized header
+		// is ahead of any children, if it is that means some previous change was not applied
+		for _, child := range root.nodes {
+			isDescendant, err := s.blockState.IsDescendantOf(child.change.announcingHeader.Hash(), finalizedHash)
+			if err != nil {
+				return nil, fmt.Errorf("cannot verify ancestry: %w", err)
+			}
+
+			if child.change.announcingHeader.Number <= finalizedNumber && isDescendant {
+				return nil, errUnfinalizedAncestor
+			}
+		}
+
+		changeNode = root
+		break
+	}
+
+	// if there is no change to be applied then we should keep only
+	// the scheduled changes which belongs to the finalized header
+	// otherwise we should update the scheduled roots to be the child
+	// nodes of the applied scheduled change
+	if changeNode == nil {
+		err := s.pruneScheduledChanges(finalizedHash)
+		if err != nil {
+			return nil, fmt.Errorf("cannot prune non-descendant scheduled nodes: %w", err)
+		}
+	} else {
+		change = changeNode.change
+		s.scheduledChangeRoots = make([]*pendingChangeNode, len(changeNode.nodes))
+		copy(s.scheduledChangeRoots, changeNode.nodes)
+	}
+
+	return change, nil
+}
+
+// pruneForcedChanges removes the forced changes from GRANDPA state for any block
+// that is not a descendant of the current finalised block.
+func (s *GrandpaState) pruneForcedChanges(finalizedHash common.Hash) error {
+	onBranchForcedChanges := make([]*pendingChange, 0, len(s.forcedChanges))
+
+	for _, forcedChange := range s.forcedChanges {
+		isDescendant, err := s.blockState.IsDescendantOf(finalizedHash, forcedChange.announcingHeader.Hash())
+		if err != nil {
+			return fmt.Errorf("cannot verify ancestry: %w", err)
+		}
+
+		if isDescendant {
+			onBranchForcedChanges = append(onBranchForcedChanges, forcedChange)
+		}
+	}
+
+	s.forcedChanges = make(orderedPendingChanges, len(onBranchForcedChanges))
+	copy(s.forcedChanges, onBranchForcedChanges)
+
+	return nil
+}
+
+// pruneScheduledChanges removes the scheduled changes from the
+// GRANDPA state which are not for a descendant of the finalised block hash.
+func (s *GrandpaState) pruneScheduledChanges(finalizedHash common.Hash) error {
+	onBranchScheduledChanges := make([]*pendingChangeNode, 0, len(s.scheduledChangeRoots))
+
+	for _, scheduledChange := range s.scheduledChangeRoots {
+		scheduledChangeHash := scheduledChange.change.announcingHeader.Hash()
+
+		isDescendant, err := s.blockState.IsDescendantOf(finalizedHash, scheduledChangeHash)
+		if err != nil {
+			return fmt.Errorf("cannot verify ancestry: %w", err)
+		}
+
+		if isDescendant {
+			onBranchScheduledChanges = append(onBranchScheduledChanges, scheduledChange)
+		}
+	}
+
+	s.scheduledChangeRoots = make([]*pendingChangeNode, len(onBranchScheduledChanges))
+	copy(s.scheduledChangeRoots, onBranchScheduledChanges)
```

Comment:
no need for the copy

```suggestion
	s.scheduledChangeRoots = onBranchScheduledChanges
```